shared/clock/clock.cpp

```diff
@@ -1,55 +1,63 @@
+    template <typename T>
+    NUClear::base_clock::duration dc(T&& t) {
+        return std::chrono::duration_cast<NUClear::base_clock::duration>(t);
+    }
 
-    void update_rtf(const double& new_rtf) {
-        auto now = NUClear::clock::now();  // Changed from base_clock
+    void update_rtf(const double& rtf) {
+        using namespace std::chrono;  // Using namespace is fine in a function scope
 
-        static std::atomic_int updateClockIndex = clockArrayIndex ^ 1;  // Switch to other clock in array
+        auto now = NUClear::base_clock::now();
 
-        // Since we are updating our rtf we need to advance our state so the new deltas will be calculated properly
-        // We set the variables in this specific order to mimise the error that will occur if the threading reads as we
-        // are writing By reducing the delta between state and now any changes in rtf will have a minimal change on
-        // delta. The old rtf is used since it is assume that since the last update the old rtf was in effect and
-        // only now with future calculations the new rtf will be used
+        // Set up a circular buffer index for the clock data
+        int c = active;
+        int n = (c + 1) % data.size();
 
-        clockArray[updateClockIndex].epoch =
-            clockArray[updateClockIndex].epoch
-            + std::chrono::duration_cast<std::chrono::steady_clock::duration>(
-                (now - clockArray[updateClockIndex].last_update)
-                * clockArray[updateClockIndex].rtf);  // set before we update the variables
-        clockArray[updateClockIndex].last_update = now;
-        clockArray[updateClockIndex].rtf         = new_rtf;
+        // Update the next clock element in the list before we set it for reading so the whole struct is complete before
+        // we let others read it
+        data[n].epoch       = data[c].epoch + dc((now - data[c].last_update) * data[c].rtf);
+        data[n].last_update = now;
+        data[n].rtf         = rtf;
 
-        clockArrayIndex ^= 1;  // Set read
+        // Set active to the clock element we just updated
+        active = n;
     }
 
 }  // namespace utility::clock
 
 namespace NUClear {
     clock::time_point clock::now() {
-        // Move along the time
-        return utility::clock::clockArray[utility::clock::clockArrayIndex].epoch
-               + std::chrono::duration_cast<std::chrono::steady_clock::duration>(
-                   (NUClear::base_clock::now()
-                    - utility::clock::clockArray[utility::clock::clockArrayIndex].last_update)
-                   * utility::clock::clockArray[utility::clock::clockArrayIndex].rtf);
+
+        using namespace utility::clock;
+        using namespace std::chrono;  // Using namespace is fine in a function scope
+
+        // Get the current index in a variable in case it changes while we use it
+        int c = active;
```

Comment:
Shouldn't this not be necessary if active is atomic? It won't update in the middle of the return line below?

Reply:
It's accessed twice in the line below, so even though it's atomic it's technically possible depending on the implementation for that to be two reads and it could change between them